vernac/vernacextend.ml

```diff
@@ -136,12 +136,14 @@ let vtcloseproof f =
   TypedVernac { inprog = Use; outprog = Yes; inproof = Use; outproof = Close;
                 run = (fun ~pm ~proof ->
                     let pm = f ~lemma:proof ~pm in
+                    Pvernac.set_current_proof_mode None;
```

Comment:
Isn't this buggy when in presence of nested proofs?

After thinking a bit more I'm unsure how the approach followed here can possibly work, without imposing a strong coupling on parsing and execution [which I understand we don't want]

I think that either we a) fix the simple compiler (and other toplevels that drive parsing) to be aware of parsing modes or b) we implement the handling of proof states for the proof mode in the parsing call itself.

As is, the PR seems a regression to me.

Reply:
I can do b, but I need a piece of info which is only available via the vernac_classifier, ie VtProofMode.
This is how the stm does it.

Would it be ok to call it just after Vernacstate.Parser.parse and return both the ast and the updated parsing state?

Anyway, I can keep this commit in a limbo until we separate parsing and execution. But I don't see the regression, since all clients still use the stm, which works as before. The simple compiler was just broken wrt proof modes, so how can it be a regression?

Reply:
The interpretation type of the command is given not by the classifier but by `Vernacxtend.typed_vernac`.

The regression is that, if I understand correctly, now we store a piece of state that is actually not correct in a few cases.

Does this PR work with the simple compiler?

You could move the parsing state setting to `interp_typed_vernac` which would fix this, however you would have to move it out again if you wanna separate parsing to execution so maybe it is more efficient to go ahead to the right solution?

How do you plan to handle this state in the separation PR?

Reply:
I can fix the case of nested proofs, it does not look hard.

Reply:
Yeah if you wanna do that and move the setting to the interpreter I think the pr is fine. Tho advanced document managers will want to hook below the vernacinterp layer (which was indeed the design), so they won't care about this too much.

Introducing a parser aware refinement of typed_vernac would be nicer IMHO.

Reply:
Please look at the last commit (I shall squash then)

Reply:
Thanks, yes please squash / edit commit messages etc... I think this should be good to go, I'll just do some testing after my talks on Fri and merge if it goes well.

What's the plan tho handle this better medium-term? I think we don't need the full parsing separation PR to add a bit of infra to handle this better and out of the exec loop (thus removing the STM hack)

Reply:
I hope the separation parsing/exec is short term, so I would not improve this further.

Reply:
done.

I'm also testing this commit in #16361

Reply:
> I hope the separation parsing/exec is short term, so I would not improve this further.

I mean you could pull the part relative to the proof mode and put in master without having the full separation in, but as you prefer.